src/bio/smd/smd_pool.c

```diff
@@ -20,267 +20,317 @@
-		  d_iov_t *val_iov, struct btr_record *rec)
-{
-	umem_off_t			 npool_off;
-	struct smd_nvme_pool_df		*npool_df;
-	struct pool_tab_key		*pkey = NULL;
-
-	D_ASSERT(key_iov->iov_len == sizeof(struct pool_tab_key));
-	pkey = (struct pool_tab_key *)key_iov->iov_buf;
-
-	npool_off = umem_zalloc(&tins->ti_umm, sizeof(struct smd_nvme_pool_df));
-	if (UMOFF_IS_NULL(npool_off))
-		return -DER_NOSPACE;
-	npool_df = umem_off2ptr(&tins->ti_umm, npool_off);
-	uuid_copy(npool_df->np_info.npi_pool_uuid, pkey->ptk_pid);
-	npool_df->np_info.npi_stream_id = pkey->ptk_sid;
-	memcpy(npool_df, val_iov->iov_buf, sizeof(struct smd_nvme_pool_df));
-	rec->rec_off = npool_off;
-
-	return 0;
-}
+	tgt_src = &entry.spe_tgts[tgt_idx + 1];
+	tgt_tgt = &entry.spe_tgts[tgt_idx];
+	blob_src = &entry.spe_blobs[tgt_idx + 1];
+	blob_tgt = &entry.spe_blobs[tgt_idx];
 
-static int
-ptab_df_rec_fetch(struct btr_instance *tins, struct btr_record *rec,
-		  d_iov_t *key_iov, d_iov_t *val_iov)
-{
-	struct smd_nvme_pool_df		*npool_df;
+	D_ASSERT(entry.spe_tgt_cnt >= (tgt_idx + 1));
+	move_cnt = entry.spe_tgt_cnt - 1 - tgt_idx;
+	entry.spe_tgt_cnt -= 1;
 
-	npool_df = umem_off2ptr(&tins->ti_umm, rec->rec_off);
-	memcpy(val_iov->iov_buf, npool_df, sizeof(struct smd_nvme_pool_df));
+	if (move_cnt && entry.spe_tgt_cnt) {
+		memmove(tgt_tgt, tgt_src, move_cnt * sizeof(int));
+		memmove(blob_tgt, blob_src, move_cnt * sizeof(uint64_t));
+	}
 
-	return 0;
+	if (entry.spe_tgt_cnt) {
+		rc = dbtree_update(smd_store.ss_pool_hdl, &key, &val);
+		if (rc)
+			D_ERROR("Update pool "DF_UUID" failed. %d\n",
+				DP_UUID(&key_pool.uuid), rc);
+	} else {
+		rc = dbtree_delete(smd_store.ss_pool_hdl, BTR_PROBE_EQ, &key,
+				   NULL);
+		if (rc)
+			D_ERROR("Delete pool "DF_UUID" failed. %d\n",
+				DP_UUID(&key_pool.uuid), rc);
+	}
+out:
+	D_RWLOCK_UNLOCK(&smd_store.ss_rwlock);
+	return rc;
 }
 
-static int
-ptab_df_rec_update(struct btr_instance *tins, struct btr_record *rec,
-		   d_iov_t *key_iov, d_iov_t *val_iov)
+static struct smd_pool_info *
+create_pool_info(uuid_t pool_id, struct smd_pool_entry *entry)
 {
-	struct smd_nvme_pool_df		*npool_df;
-	int				 rc;
-
-	rc = umem_tx_add(&tins->ti_umm, rec->rec_off,
-			 sizeof(struct smd_nvme_pool_df));
-	if (rc != 0)
-		return rc;
+	struct smd_pool_info	*info;
+	int			 i;
 
-	npool_df = umem_off2ptr(&tins->ti_umm, rec->rec_off);
-	memcpy(npool_df, val_iov->iov_buf, val_iov->iov_len);
-	return 0;
-}
+	D_ALLOC_PTR(info);
+	if (info == NULL)
+		return NULL;
 
-btr_ops_t ptab_ops = {
-	.to_hkey_size	= ptab_df_hkey_size,
-	.to_hkey_gen	= ptab_df_hkey_gen,
-	.to_hkey_cmp	= ptab_df_hkey_cmp,
-	.to_rec_alloc	= ptab_df_rec_alloc,
-	.to_rec_free	= ptab_df_rec_free,
-	.to_rec_fetch	= ptab_df_rec_fetch,
-	.to_rec_update	= ptab_df_rec_update,
-};
-
-int
-smd_nvme_ptab_register()
-{
-	int	rc;
-
-	D_DEBUG(DB_DF, "Register peristent metadata pool index: %d\n",
-		DBTREE_CLASS_SMD_PTAB);
+	D_ALLOC_ARRAY(info->spi_tgts, SMD_MAX_TGT_CNT);
```

Comment:
I would prefer to keep them as static arrays this would avoid additional allocations

Reply:
That'll expose the internal implementation limitation which I don't quite like.